pkg/plugins/runtime/k8s/controllers/gatewayapi/attachment/attachment.go

```diff
@@ -0,0 +1,181 @@
+	Invalid
+	// Attached means we can successfully attached
+	Attached
+)
+
+func listenerReady(
+	statuses []gatewayapi.ListenerStatus,
+	name gatewayapi.SectionName,
+) bool {
+	for _, status := range statuses {
+		if status.Name != name {
+			continue
+		}
+
+		return kube_apimeta.IsStatusConditionTrue(status.Conditions, string(gatewayapi.ListenerConditionReady))
+	}
+
+	return false
+}
+
+// findRouteListenerAttachment reports whether this ref is allowed to attach to
+// the Gateway or specific Listener.
+func findRouteListenerAttachment(
+	gateway *gatewayapi.Gateway,
+	routeNs kube_client.Object,
+	refSectionName *gatewayapi.SectionName,
+) (Attachment, error) {
+	// Build a map of whether attaching to each listener is possible
+	listeners := map[gatewayapi.SectionName]Attachment{}
+
+	for _, l := range gateway.Spec.Listeners {
+		ns := l.AllowedRoutes.Namespaces
+
+		// TODO
+		// The gateway controller ensures Kinds is either empty or contains
+		// HTTPRoute, so we can ignore it here.
+
+		switch *ns.From {
+		case gatewayapi.NamespacesFromSelector:
+			// TODO, the gateway controller/webhook should verify this
+			selector, err := kube_meta.LabelSelectorAsSelector(ns.Selector)
+			if err != nil {
+				return Unknown, errors.Wrap(err, "internal error: couldn't convert to selector")
+			}
+
+			if !selector.Matches(kube_labels.Set(routeNs.GetLabels())) {
+				listeners[l.Name] = NotPermitted
+				continue
+			}
+		case gatewayapi.NamespacesFromSame:
+			if gateway.Namespace != routeNs.GetName() {
+				listeners[l.Name] = NotPermitted
+				continue
+			}
+		case gatewayapi.NamespacesFromAll:
+		}
+
+		if listenerReady(gateway.Status.Listeners, l.Name) {
+			listeners[l.Name] = Attached
+		} else {
+			listeners[l.Name] = Unknown
+		}
+	}
+
+	sectionName := ""
+	if refSectionName != nil {
+		sectionName = string(*refSectionName)
+	}
+
+	// Look through the potential Listeners:
+	// If it's our specific listener, we return that status
+	for name, status := range listeners {
+		if string(name) == sectionName {
+			return status, nil
+		}
+	}
+
+	// If we don't find our listener, our ref is invalid
+	if sectionName != "" {
+		return Invalid, nil
+	}
+
+	// If we aren't attaching to a specific listener then
+	// as soon as one attaches we're attached
+	for _, status := range listeners {
+		if status == Attached {
+			return Attached, nil
+		}
+	}
+
+	return NotPermitted, nil
+}
+
+// getParentRefGateway returns a kuma-class Gateway if one was found,
+// otherwise it returns nil.
+func getParentRefGateway(
+	ctx context.Context,
+	client kube_client.Client,
+	fromNamespace string,
+	ref gatewayapi.ParentRef,
+) (*gatewayapi.Gateway, error) {
+	name := string(ref.Name)
+	// Group and Kind both have default values
+	group := string(*ref.Group)
+	kind := string(*ref.Kind)
+
+	namespace := fromNamespace
+	if ns := ref.Namespace; ns != nil {
+		namespace = string(*ns)
+	}
+
+	if group != gatewayapi.GroupName || kind != "Gateway" {
+		return nil, nil
```

Comment:
Yeah of course right now, there's only `Gateway` support so it's probably a mistake. But these are the defaults, so the user likely won't even set them explicitly, which makes a typo somewhat unlikely.

Since a route doesn't belong to a `GatewayClass`, we should just ignore this, it's not our controller's responsibility to validate routes unless we know they refer to one of our `Gateway`s. I suppose a log message can't hurt.